pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,79 +22,309 @@ import (
+	podStoreSynced func() bool
+
+	// Deployments that need to be synced
+	queue *workqueue.Type
 }
 
-func New(client client.Interface) *DeploymentController {
+// NewDeploymentController creates a new DeploymentController.
+func NewDeploymentController(client client.Interface, resyncPeriod controller.ResyncPeriodFunc) *DeploymentController {
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(glog.Infof)
 	eventBroadcaster.StartRecordingToSink(client.Events(""))
 
-	return &DeploymentController{
+	dc := &DeploymentController{
 		client:        client,
 		expClient:     client.Extensions(),
 		eventRecorder: eventBroadcaster.NewRecorder(api.EventSource{Component: "deployment-controller"}),
+		queue:         workqueue.New(),
 	}
+
+	dc.dStore.Store, dc.dController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.expClient.Deployments(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.expClient.Deployments(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&extensions.Deployment{},
+		FullDeploymentResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: dc.enqueueDeployment,
+			UpdateFunc: func(old, cur interface{}) {
+				// Resync on deployment object relist.
+				dc.enqueueDeployment(cur)
+			},
+			// This will enter the sync loop and no-op, because the deployment has been deleted from the store.
+			// Note that deleting a controller immediately after scaling it to 0 will not work.
+			DeleteFunc: dc.enqueueDeployment,
+		},
+	)
+
+	dc.rcStore.Store, dc.rcController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.client.ReplicationControllers(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.client.ReplicationControllers(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.ReplicationController{},
+		resyncPeriod(),
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dc.addRC,
+			UpdateFunc: dc.updateRC,
+			DeleteFunc: dc.deleteRC,
+		},
+	)
+
+	// We do not event on anything from the podController, but we use the local
+	// podStore to make queries about the current state of pods (e.g. whether
+	// they are ready or not) more efficient.
+	dc.podStore.Store, dc.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.client.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.client.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.Pod{},
+		podRelistPeriod,
```

Comment:
I am wondering if this is because we are not doing anything on pod updates.
Consider the situation where I create a deployment with 3 replicas (there are no existing pods).
In first run of syncDeployment, we will create an RC and scale it to 3. deployment.updatedReplicas will be 0 (since none of those pods have become ready yet).
Now, no one will update RC and hence deployment will never by enqueued again, right?

We can fix this by enquing the deployment again when the pod updates (becomes ready).